diff -uNr a/Makefile b/Makefile
--- a/Makefile	2005-08-08 10:18:41.000000000 +1000
+++ b/Makefile	2016-10-02 22:53:41.000000000 +1100
@@ -231,8 +231,8 @@
 endif
 endif
 
-CFLAGS_RELEASE := -Iinclude $(WFLAGS) $(OFLAGS)
-CFLAGS_DEBUG := -Iinclude $(WFLAGS) $(DBGFLAGS)
+CFLAGS_RELEASE := -Iinclude $(WFLAGS) $(OFLAGS) $(PLAT_CFLAGS)
+CFLAGS_DEBUG := -Iinclude $(WFLAGS) $(DBGFLAGS) $(PLAT_CFLAGS)
 
 LDFLAGS := -s
 
@@ -349,6 +349,7 @@
 veryclean: clean
 	$(call DELETE,$(CORE_LIB_FILE))
 	$(call DELETE,$(ALLEGRO_LIB_FILE))
+	$(call DELETE,$(LIBDIR)/*.a)
 	$(call DELETE,$(CORE_EXAMPLES_EXE))
 	$(call DELETE,$(ALLEGRO_EXAMPLES_EXE))
 
diff -uNr a/Makefile.rdy b/Makefile.rdy
--- a/Makefile.rdy	2005-08-08 10:18:39.000000000 +1000
+++ b/Makefile.rdy	2016-10-02 22:33:06.000000000 +1100
@@ -231,8 +231,8 @@
 endif
 endif
 
-CFLAGS_RELEASE := -Iinclude $(WFLAGS) $(OFLAGS)
-CFLAGS_DEBUG := -Iinclude $(WFLAGS) $(DBGFLAGS)
+CFLAGS_RELEASE := -Iinclude $(WFLAGS) $(OFLAGS) $(PLAT_CFLAGS)
+CFLAGS_DEBUG := -Iinclude $(WFLAGS) $(DBGFLAGS) $(PLAT_CFLAGS)
 
 LDFLAGS := -s
 
diff -uNr a/include/dumb.h b/include/dumb.h
--- a/include/dumb.h	2005-08-08 10:18:40.000000000 +1000
+++ b/include/dumb.h	2016-10-02 22:33:18.000000000 +1100
@@ -38,7 +38,7 @@
 
 #define DUMB_VERSION_STR "0.9.3"
 
-#define DUMB_NAME "DUMB v"DUMB_VERSION_STR
+#define DUMB_NAME "DUMB v" DUMB_VERSION_STR
 
 #define DUMB_YEAR  2005
 #define DUMB_MONTH 8
@@ -50,13 +50,13 @@
 #define DUMB_DAY_STR1   "7"
 
 #if DUMB_MONTH < 10
-#define DUMB_MONTH_STR2 "0"DUMB_MONTH_STR1
+#define DUMB_MONTH_STR2 "0" DUMB_MONTH_STR1
 #else
 #define DUMB_MONTH_STR2 DUMB_MONTH_STR1
 #endif
 
 #if DUMB_DAY < 10
-#define DUMB_DAY_STR2 "0"DUMB_DAY_STR1
+#define DUMB_DAY_STR2 "0" DUMB_DAY_STR1
 #else
 #define DUMB_DAY_STR2 DUMB_DAY_STR1
 #endif
@@ -68,7 +68,7 @@
  */
 #define DUMB_DATE (DUMB_YEAR*10000 + DUMB_MONTH*100 + DUMB_DAY)
 
-#define DUMB_DATE_STR DUMB_DAY_STR1"."DUMB_MONTH_STR1"."DUMB_YEAR_STR4
+#define DUMB_DATE_STR DUMB_DAY_STR1 "." DUMB_MONTH_STR1 "." DUMB_YEAR_STR4
 
 
 #undef MIN
