diff -Naur make/ios.inc make/ios.inc
--- make/ios.inc	1970-01-01 01:00:00.000000000 +0100
+++ make/ios.inc	2011-10-11 10:09:49.070435600 +0200
@@ -0,0 +1,21 @@
+# This file contains definitions suitable for iOS systems. It will
+# be included by the main Makefile when you compile on such a system.
+
+PLATFORM := ios
+
+APOST := \'
+
+# Macro that on DOS and Windows would replace / with \. Usage: $(call FIX,path)
+FIX = $(1)
+
+ECHO = @echo $(1)
+DELETE = rm -f $(1)
+COPY = cp $(1) $(2)
+MKDIR = mkdir $(1)
+
+EXE_SUFFIX :=
+
+LINK_MATH := -lm
+LINK_ALLEGRO :=
+
+# PREFIX is set by config.sh.
\ No newline at end of file
diff -Naur make/config.txt make/config.txt
--- make/config.txt	1970-01-01 01:00:00.000000000 +0100
+++ make/config.txt	2011-10-11 10:18:05.885523100 +0200
@@ -0,0 +1,3 @@
+include make/ios.inc
+ALL_TARGETS := core core-headers
+ALL_TARGETS += allegro allegro-headers
diff -Naur Makefile Makefile
--- Makefile	2005-08-08 02:18:41.000000000 +0200
+++ Makefile	2011-10-19 08:57:16.416901400 +0200
@@ -41,10 +41,10 @@
 ifdef USE_SGICC
 CC := cc
 else
-CC := gcc
+CC := $(shell xcrun --sdk iphonesimulator --find gcc)
 endif
 endif
-AR := ar
+AR := $(shell xcrun --sdk iphonesimulator --find ar)


 # Configuration.
@@ -74,6 +74,7 @@

 else

+PREFIX := $(IOS_ADDITIONAL_LIBRARY_PATH)
 ifdef PREFIX
 DEFAULT_PREFIX := $(PREFIX)
 else
@@ -231,8 +232,10 @@
 endif
 endif

-CFLAGS_RELEASE := -Iinclude $(WFLAGS) $(OFLAGS)
-CFLAGS_DEBUG := -Iinclude $(WFLAGS) $(DBGFLAGS)
+CFLAGS_IOS = $(CFLAGS)
+
+CFLAGS_RELEASE := -Iinclude $(WFLAGS) $(OFLAGS) $(CFLAGS_IOS)
+CFLAGS_DEBUG := -Iinclude $(WFLAGS) $(DBGFLAGS) $(CFLAGS_IOS)

 LDFLAGS := -s

diff -Naur Makefile.rdy Makefile.rdy
--- Makefile.rdy	2005-08-08 02:18:39.000000000 +0200
+++ Makefile.rdy	2011-10-19 08:57:16.416901400 +0200
@@ -41,10 +41,10 @@
 ifdef USE_SGICC
 CC := cc
 else
-CC := gcc
+CC := $(shell xcrun --sdk iphonesimulator --find gcc)
 endif
 endif
-AR := ar
+AR := $(shell xcrun --sdk iphonesimulator --find ar)


 # Configuration.
@@ -74,6 +74,7 @@

 else

+PREFIX := $(IOS_ADDITIONAL_LIBRARY_PATH)
 ifdef PREFIX
 DEFAULT_PREFIX := $(PREFIX)
 else
@@ -231,8 +232,10 @@
 endif
 endif

-CFLAGS_RELEASE := -Iinclude $(WFLAGS) $(OFLAGS)
-CFLAGS_DEBUG := -Iinclude $(WFLAGS) $(DBGFLAGS)
+CFLAGS_IOS = $(CFLAGS)
+
+CFLAGS_RELEASE := -Iinclude $(WFLAGS) $(OFLAGS) $(CFLAGS_IOS)
+CFLAGS_DEBUG := -Iinclude $(WFLAGS) $(DBGFLAGS) $(CFLAGS_IOS)

 LDFLAGS := -s

